File: 프로그래머스/KaKao/거리두기 확인하기.py
# 2021 카카오 채용연계형 인턴십 거리두기 확인하기
flag = True
dir = [[1, 0], [-1, 0], [0, 1], [0, -1]]


def DFS(place, visited, y, x, level):
    global flag

    if not flag:
        return

    if level >= 2:
        flag = True
        return

    for d in dir:
        ny = y + d[0]
        nx = x + d[1]

        # 범위 내
        if ny >= 0 and ny < 5 and nx >= 0 and nx < 5:
            if place[ny][nx] == "X" or visited[ny][nx]:
                continue

            if level < 2 and place[ny][nx] == "P":
                flag = False
                return

            if place[ny][nx] == "O":
                visited[ny][nx] = True
                DFS(place, visited, ny, nx, level+1)
    return


def checkPlace(place):
    global flag
    flag = True
    visited = [[False] * 5 for _ in range(5)]

    for i in range(5):
        for j in range(5):
            # 사람 발견
            if place[i][j] == "P":
                visited[i][j] = True
                DFS(place, visited, i, j, 0)

            if not flag:
                return 0

    return 1 if flag else 0


def solution(places):
    answer = []

    for place in places:
        answer.append(checkPlace(place))
    return answer

File: 프로그래머스/KaKao/거리두기 확인하기.cpp
#include <string>
#include <vector>
#include <cstring>

using namespace std;

bool flag = true;
int visited[5][5];
int dir[4][2] = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

void DFS(vector<string> place, int x, int y, int depth) {
    if (!flag) // 불통이 나왔다면
        return;
    
    if (depth == 2) {
        return;
    }
    
    for (auto &d: dir) {
        int nx = x + d[0];
        int ny = y + d[1];
        
        if (nx < 0 || ny < 0 || nx >= 5 || ny >= 5) continue;
        
        if (visited[nx][ny]) continue;
        
        if (place[nx][ny] == 'X') continue;
        
        if (depth < 2 && place[nx][ny] == 'P') {
            flag = false;
            return;
        }
        
        visited[nx][ny] = 1;
        DFS(place, nx, ny, depth+1);
        visited[nx][ny] = 0;
    }
    
    return;
}

bool checkPlaces(vector<string> place) {
    flag = true;
    for (int i = 0; i < 5; ++i) {
        for (int j = 0; j < 5; ++j) 
        {
            if (place[i][j] == 'P') 
            {
                memset(visited, 0, sizeof(visited));
                visited[i][j] = 1;
                DFS(place, i, j, 0);
            }
        }
    }
    
    return flag;
}

vector<int> solution(vector<vector<string>> places) {
    vector<int> answer;
    
    for (auto &place: places) {
        int check = checkPlaces(place);
        answer.push_back(check);
    }
    return answer;
}